testing/integration/otel_test.go

```diff
@@ -0,0 +1,288 @@
+	info := define.Require(t, define.Requirements{
+		Group: Default,
+		Stack: &define.Stack{},
+		Local: true,
+		OS: []define.OS{
+			// apm server not supported on darwin
+			{Type: define.Linux},
+		},
+	})
+
+	const apmVersionMismatch = "The APM integration must be upgraded"
+	const apmReadyLog = "all precondition checks are now satisfied"
+	logWatcher := aTesting.NewLogWatcher(t,
+		apmVersionMismatch, // apm version mismatch
+		apmReadyLog,        // apm ready
+	)
+
+	// prepare agent
+	fixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	ctx, cancel := testcontext.WithDeadline(t, context.Background(), time.Now().Add(10*time.Minute))
+	defer cancel()
+	err = fixture.Prepare(ctx, fakeComponent, fakeShipper)
+	require.NoError(t, err)
+
+	// prepare input
+	agentWorkDir := fixture.WorkDir()
+	fileName := "content.log"
+	fixture.WriteFileToWorkDir(ctx, "", fileName)
+
+	testId := info.Namespace
+
+	apmConfig := fmt.Sprintf(apmOtelConfig, filepath.Join(agentWorkDir, fileName), testId)
+
+	err = fixture.Configure(ctx, []byte(apmConfig))
+	require.NoError(t, err)
+
+	componentsDir, err := aTesting.FindComponentsDir(agentWorkDir)
+	require.NoError(t, err)
+
+	// start apm default config just configure ES output
+	esHost, err := getESHost()
+	require.NoError(t, err, "failed to get ES host")
+	require.True(t, len(esHost) > 0)
+
+	esClient := info.ESClient
+	esApiKey, err := createESApiKey(esClient)
+	require.NoError(t, err, "failed to get api key")
+	require.True(t, len(esApiKey) > 1, "api key is invalid %q", esApiKey)
+
+	apmArgs := []string{
+		"run",
+		"-e",
+		"-E", "output.elasticsearch.hosts=['" + esHost + "']",
+		"-E", "output.elasticsearch.api_key=" + esApiKey,
+		"-E", "apm-server.host=127.0.0.1:8200",
+		"-E", "apm-server.ssl.enabled=false",
+	}
+
+	apmPath := filepath.Join(componentsDir, "apm-server")
+	var apmFixtureWg sync.WaitGroup
+	apmFixtureWg.Add(1)
+	apmContext, apmCancel := context.WithCancel(ctx)
+	defer apmCancel()
+	go func() {
+		aTesting.RunProcess(t,
+			logWatcher,
+			apmContext, 0,
+			true, true,
+			apmPath, apmArgs...)
+		apmFixtureWg.Done()
+	}()
+
+	// start agent
+	var fixtureWg sync.WaitGroup
+	fixtureWg.Add(1)
+	go func() {
+		fixture.RunWithClient(ctx, false)
+		fixtureWg.Done()
+	}()
+
+	err = logWatcher.WaitForKeys(context.Background(),
+		10*time.Minute,
+		500*time.Millisecond,
+		apmReadyLog,
+	)
+	require.NoError(t, err, "APM not initialized")
+	fixture.WriteFileToWorkDir(ctx, apmProcessingContent, fileName)
+
+	// check index
+	var hits int
+	match := map[string]interface{}{
+		"labels.host_test-id": testId,
+	}
+
+	// apm mismatch or proper docs in ES
+	err = logWatcher.WaitForKeys(context.Background(),
+		20*time.Second,
+		500*time.Millisecond,
+		apmVersionMismatch)
+	if err == nil {
+		// err nil means there's a match within time limit
+		// mark skipped to make it explicit it was not successfully evaluated
+		t.Skip("agent version needs to be equal to stack version")
```

Comment:
Why can we only make this check at the end? Don't we know the versions we are starting at the start of the test?

Reply:
Also why is this check timing dependent? How would you catch cases where this test is being skipped when it shouldn't be?

Reply:
we cannot check the version because we know version of agent and version of stack. what's causing issues is version of APM integration. This bubbles up only during runtime as an APM log.

Reply:
You could probably use the Fleet API to get the version of the APM integration instead. Probably the package policies API to find the APM package details https://www.elastic.co/guide/en/fleet/8.11/fleet-apis.html#PackagePolicies